certification/internal/shell/has_v2_related_image_schema_version.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/errors"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/cli"
+	log "github.com/sirupsen/logrus"
+)
+
+// RelatedImagesAreSchemaVersion2Check is part of the Operator policy and implements
+// the Check interface.
+type RelatedImagesAreSchemaVersion2Check struct{}
+
+// Validate checks the image manifest for each related image referenced in a
+// ClusterServiceVersion and ensures that the schema version used is version 2.
+func (p *RelatedImagesAreSchemaVersion2Check) Validate(image string) (bool, error) {
+
+	imageToSchemaVersion, err := p.getDataToValidate(image)
+	if err != nil {
+		return false, fmt.Errorf("%w: %s", errors.ErrRunContainerFailed, err)
+	}
+
+	return p.validate(imageToSchemaVersion)
+}
+
+// getDataToValidate pulls a ClusterServiceVersion from the input operator bundle,
+// checks the ClusterServiceVersion's related images declaration, and assembles a list of
+// images and their image manifest schema version values.
+func (p *RelatedImagesAreSchemaVersion2Check) getDataToValidate(image string) (map[string]int, error) {
+	bundlePath, err := p.extractBundleFromImage(image)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+	csv, err := p.readBundle(bundlePath)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+
+	relatedImages, err := p.getRelatedImagesForCSV(csv)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+
+	imageSchemaVersionMap, err := p.inspectSchemaVersionForImage(relatedImages)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+
+	return imageSchemaVersionMap, nil
+}
+
+// extractBundleFromImage will extract the bundle contents from the specified image, and return the
+// path to the bundle on the filesystem. This is done by creating the container, copying the contents
+// out of the container, and then cleaning up the container.
+func (p *RelatedImagesAreSchemaVersion2Check) extractBundleFromImage(image string) (string, error) {
+	localBundlePath := "extracted-bundle"
+
+	// create the container. Bundle containers are non-runnable, so we need to
+	// add a dummy entrypoint so that we can create the container.
+	createReport, err := podmanEngine.Create(image, &cli.PodmanCreateOptions{Entrypoint: "false"})
+	if err != nil {
+		log.Error("stdout: ", createReport.Stdout)
+		log.Error("stderr: ", createReport.Stderr)
+		return "", err
+	}
+
+	copyReport, err := podmanEngine.CopyFrom(createReport.ContainerID, "/", localBundlePath)
+	if err != nil {
+		log.Error("stdout: ", copyReport.Stdout)
+		log.Error("stderr: ", copyReport.Stderr)
+		return "", err
+	}
+
+	deleteReport, err := podmanEngine.Remove(createReport.ContainerID)
```

Comment:
How do you feel about changing this to defer podmanEngine.Remove() and ignoring the return? We'd want the container cleaned up if CopyFrom fails.

Reply:
@bcrochet Hmm, great point. I changed this defer the removal, and take any errors that return from deleting the container into warning log lines so that the log should reflect it at the warn log level. Check it, let me know if that works for you!